### ISSUE TITLE
App size too inflated?

### ISSUE THREAD

**Issue by sonora** [osmandapp/OsmAnd#14336]
I have received several requests now asking if it was on purpose that the app (apk) size has increased by some 30% or so between 4.0.x and higher versions.

I remember we added some AndroidAuto code, and raised our minSDK requirement, but I have to admit that I am a bit surprised myself that the app size increased by 20 or 30 MB?

**Comment by pebogufi** [comment]
Some details:
4.1.0#33144.  28/11/2021 123,6MB
4.1.0#33442.  21/01/2022 124,2MB
4.1.0#33465.  25/01/2022 128,3MB

Sorry, I don't have older data.

**Comment by sonora** [comment]
4.0.9 was around 94MB, see https://download.osmand.net/releases/net.osmand-4.0.9.apk

**Comment by vshcherb** [comment]
It was expected cause in 4.2 it bundles with opengl engine now.

**Comment by vshcherb** [comment]
https://download.osmand.net/latest-night-build/OsmAnd-opengl-arm64-nightly.apk it's by 50 MB bigger than Android

**Comment by vshcherb** [comment]
Also mini basemap is now + 10 MB - https://builder.osmand.net/basemap/release/2022-01/

**Comment by vshcherb** [comment]
Most significant changes
```
OsmAnd-default-nb-2021-09-13.apk                   13-Sep-2021 21:27           105993422
OsmAnd-default-nb-2021-09-20.apk                   20-Sep-2021 13:02           127619948

OsmAnd-default-nb-2022-01-24.apk                   24-Jan-2022 17:57           130327835
OsmAnd-default-nb-2022-01-28.apk                   28-Jan-2022 17:57           134575258
```

**Comment by sonora** [comment]
But our default builds are not at all open-gl capable, or are they?

And still, their file size increased from 105 to 135 in recent months?

**Comment by sonora** [comment]
September 2021 is only 8 months ago ... ;)

But to the poinit: If you look at our releases at https://download.osmand.net/releases/, the jump happened between v4.09 in September 2021 and v4.1.1 in December 2021. I was just wondering what the respective increase in functionality was, AA, mini-basemap? Dont't think it was open-gl because these builds don't support it, or am I wrong?

BTW: Users ask me if they should manually de-activate the mini-basemap once they have downloaded the real basemap to avoid the renderer to double-process?

**Comment by vshcherb** [comment]
I think you are right nightly includes only:
- Android Auto 
- Mini Basemap

**Comment by sonora** [comment]
Yes, valid point.

Regarding basemap mini, I guess what triggered users is that under "local maps" we show both maps as active, so if mini is automatically disabled and we don't reflect this, this could be a bug:

![Screenshot_20220601-223947_OsmAnd~](https://user-images.githubusercontent.com/845735/171497901-29a7fe93-f195-418e-a1f5-7f50b45b2045.jpg)

**Comment by vshcherb** [comment]
@sonora we have ongoing same conversations about TTS always copied or files dispalyed and not active. So we need to choose 
1) Either display all files present in the folder
2) Start filtering files by some unknown yet strategy. Strategy could be different for different group of files

**Comment by sonora** [comment]
To implement more 'hygiene' here, could perhaps move the 'mini' file to the backup folder once user downloads the full basemap, then our code would automatically show it it as 'deactivated'?

General update strategy, also for  TTS files could be: Place all files prresent in the apk itself in the backup folder (overwriting any predecessors there), except a previous version is present in the active-maps or active-voice folder, in which case we would overwrite it there? This strategy would ensure that any data files deactivated by the user would remain deactivated even if the app is updated and the apk contains some updated data files?

**Comment by vshcherb** [comment]
Assets should be decreased obviously:
- 71M OsmAndCore_ResourcesBundle (35 MB zip) - not needed at all
- 12M fonts (6.4 MB zip) - download later?
- 45M World_basemap_mini.obf (28.6 MB zip) - make smaller? download later?
- 2.3M Default_wikivoyage.travel.obf (1.8 MB zip)

**Comment by vshcherb** [comment]
As a side comment for world basemap mini:
Tiles of 1-6 zoom - 53 MB
Tiles of 1-7 zoom - 150 MB
World basemap mini - 30 MB 
World basemap - 267 MB

**Comment by ivanPyrohivskyi** [comment]
/assets/OsmAndCore_ResourcesBundle

Can be optimized:
map/fonts - 6.1Mb, need add external initializating in EmbeddedTypefaceFinder (Android/iOS)
map/styles/default.render.xml.gz - 88Kb, need add external initializating in ResourcesManager (Android/iOS)
routing/routing.xml.qz - 22Kb, need external initializating in RoutingConfiguration (iOS only?)

Difficult to optimize:
[ddf=x.x]/map/icons
[ddf=x.x]/map/shields
[ddf=x.x]/map/shaders
[ddf=x.x]/map/stubs
Total size 10.6Mb.
For Android need to create special provider that provide icons from Android (throw Swig) to core library.
In Android for legacy code present method getIconRawData in RenderingContext (RenderingIcons) class.
And divide icons initialization iOS and Android.

Can not be optimized:
misc/icu4c/icu-data-l.dat.gz - 7.1Mb, using in core library only

Technical moment:
All resources for OsmAndCore_ResourcesBundle describe in core/warappers/android/build.gradle

**Comment by sonora** [comment]
I just notice that the analysis in @vshcherb 's  picture of December 10 above compares v4.1.1 with v4.2.4, which in deed shows a big increase. But there was an additional not plausible increase of the order of 30%(!) immediately before that from v4.0.9 to v4.1.1. Maybe including that in the above analysis picture would give additional hints?

In general, I think our 'First app start' handling has progressed enough lately so we could place some tasks there, like not only recommending to download the regional full map for the user's location, but also offering to download a mini or full basemap, plus perhaps special fonts. Then these could be taken out of the app. Their maintenance cycle is separate from the app's, anyway.

**Comment by vshcherb** [comment]
@sonora I don't see any increase except releasing opengl version
![Screenshot 2023-02-27 at 13 03 01](https://user-images.githubusercontent.com/1042025/221547174-57d34649-07f7-45a2-b913-3e44351d78a6.png)

**Comment by sonora** [comment]
Here is our release apk size development over time, sorted by time: [apksizes-over-time.txt](https://github.com/osmandapp/OsmAnd/files/10839561/apksizes-over-time.txt) (as taken from https://download.osmand.net/releases/).


The bump I talk about is in line 436:

![image](https://user-images.githubusercontent.com/845735/221563311-27b71561-a366-46bb-b997-7552bbe493cf.png)

**Comment by sonora** [comment]
I think the apks listed there are all "fat" builds...

**Comment by sonora** [comment]
So what do you think, shall we at least move the mini-basemap and the special fonts out of the apk to separate in-app downloads, plus offer them on our 'First app start' screen?

**Comment by vshcherb** [comment]
Nope, that's non-sense it only saves disk space for updates which is not needed cause google play already does it automatically. Solution is only
1. Reduce size of mini basemap
2. Double check that fonts are download with specific maps not for all apps if possible
3. Reduce resources usage by letting opengl engine reuse vector drawables from android resources

It could save up to 50-60 MB

**Comment by sonora** [comment]
My thinking is that as long as we include it in the apk, even if users delete the basemap or special fonts, this dara will continue to occupy system storage like everything which is part of an apk. If it was a separate download (like that standard basemap is), it will not consume system storage, but only internal or external storage (wherever your OsmAnd data folder is), and only while present.

**Comment by Arian03dc** [comment]
> My thinking is that as long as we include it in the apk, even if users delete the basemap or special fonts, this data will continue to occupy system storage like everything which is part of an apk. If instead it was a separate download (like that standard basemap is), it will not consume system storage, but only internal or external storage (wherever your OsmAnd data folder is), and only while present.

Good idea, by example, I don't want world basemap, only Cuba map, and the fonts files can be deleted if isn't neccesary

**Comment by sonora** [comment]
So whatc do we think, can I re-table the following idea for perhaps our v4.5 milestone?

> ...shall we at least move the mini-basemap and the special fonts out of the apk to separate in-app downloads, plus offer them on our 'First app start' screen?

Perhaps just for the basemap mini, which few people may actively use as a lasting map, but it (if part of the apk) remains to reside in your internal device memory forever even if user-disabled in the app or user-updated to the full basemap?

**Comment by sonora** [comment]
Quite possibly we have two types of users:
- the ones who will download a full basemap anyway, from which point on them having a basemap-mini is dead freight
- and the ones who never care about  basemap using OsmAnd purely on a regional level.
Would our telemetry-data give any insight into how the actual numbers are?

**Comment by vshcherb** [comment]
With downloads it's still 10% download full map but it's not perfect data cause iOS requires to download full map we still need to reiterate

**Comment by sonora** [comment]
I have tested that excluding the basemap_mini from the apk
- reduces the apk size by 30MB (for all flavors), 
- and updating with that apk permanently reduces the "App storage" Android reports as "used by OsmAnd in Internal storage" also by 30 MB. (This Internal storage saving can only be achieved if the basemap mini is not part pf the apk - if it is, it will after aop installation pemanently occupy 30 MB here even after in the app the user deletes the basemap_mini, e.g. after downloading a full basemap).
